src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MatchXpathCheck.java

```diff
@@ -0,0 +1,148 @@
+import net.sf.saxon.trans.XPathException;
+
+/**
+ * <p>
+ * Evaluates Xpath query and report violation on matching AST nodes. All nodes
+ * retrieved by Xpath evaluator will be logged as violations. Current check allows
+ * user to implement custom checks using Xpath.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code query} - Xpath query.
+ * Type is {@code java.lang.String}.
+ * Default value is {@code ""}.
+ * </li>
+ * </ul>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="MatchXpath"/&gt;
+ * </pre>
+ * <p>
+ * To fail with violation if method name is 'test' or 'foo'
+ * </p>
+ * <pre>
+ * &lt;module name="MatchXpath"&gt; &lt;property name="query"
+ * value="//METHOD_DEF[./IDENT[@text='test' or @text='foo']]"/&gt; &lt;/module&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * public class Test {
+ *  public void test() {} // violation
+ *  public void getName() {} // ok
+ *  public void foo() {} // violation
+ *  public void sayHello() {} // ok
+ * }
+ * </pre>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code matchxpath.match}
+ * </li>
+ * </ul>
+ *
+ * @since 8.37
+ */
+@StatelessCheck
+public class MatchXpathCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text provided by user.
+     */
+    public static final String MSG_KEY = "matchxpath.match";
+
+    /** Xpath query. */
+    private String query = "";
+
+    @Override
+    public void beginTree(DetailAST rootAST) {
+        if (!query.isEmpty()) {
+            evaluateXpathQuery(rootAST);
+        }
+    }
+
+    private void evaluateXpathQuery(DetailAST rootAST) {
+        final XPathEvaluator xpathEvaluator = new XPathEvaluator(Configuration.newConfiguration());
+        try {
+            final RootNode rootNode = new RootNode(rootAST);
+            final XPathExpression xpathExpression = xpathEvaluator.createExpression(query);
+            final XPathDynamicContext xpathDynamicContext =
+                    xpathExpression.createDynamicContext(rootNode);
+            final List<Item> matchingItems = xpathExpression.evaluate(xpathDynamicContext);
+
+            matchingItems.forEach(node -> {
+                final AbstractNode xpathNode = (AbstractNode) node;
+                log(xpathNode.getUnderlyingNode(), MSG_KEY);
+            });
+        }
+        catch (XPathException ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return CommonUtil.EMPTY_INT_ARRAY;
+    }
+
+    /**
+     * Setter to xpath query.
+     *
+     * @param query xpath query.
+     */
+    public void setQuery(String query) {
```

Comment:
move this setter on top of methods.

then should be all getXXXxxxToken

then should be beginTree

Reply:
done